contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -2689,6 +2699,169 @@ pltsql_detect_numeric_overflow(int weight, int dscale, int first_block, int nume
 	return (total_digit_count > TDS_NUMERIC_MAX_PRECISION);
 }
 
+/*
+ * Throw a common error message while casting to time datatype
+ */
+static void gen_func_time_in_error()
+{
+	ereport(ERROR,
+			(errcode(ERRCODE_INVALID_DATETIME_FORMAT),
+			errmsg("Conversion failed when converting date and/or time from character string.")));
+}
+
+static bool containsInMonthFormat(int *ftype){
+	int count0 = 0, count1 = 0;
+	for(int i=0;i<3;i++)
+	{
+		if(ftype[i] == DTK_NUMBER)
+			count0++;
+		else if(ftype[i] == DTK_STRING)
+			/* To-Do Check if valid text month is given */
+			count1++;
+		else
+			return false;
+	}
+	if(count0 ==2 && count1 ==1)
+		return true;
+	return false;
+}
+
+Datum pltsql_time_in(const char *str, int32	typmod)
+{
+	TimeADT		result;
+	fsec_t		fsec;
+	struct pg_tm tt,
+			   *tm = &tt;
+	int			tz;
+	int			nf;
+	int			dterr;
+	char		workbuf[MAXDATELEN + 1];
+	char	   *field[MAXDATEFIELDS];
+	int			dtype;
+	int			ftype[MAXDATEFIELDS];
+	StringInfo	sf_1 ;
+	int status;
+
+	dterr = ParseDateTime(str, workbuf, sizeof(workbuf),
+						field, ftype, MAXDATEFIELDS, &nf);
+	if (dterr == 0)
+	{
+		if(nf >= 3)
+		{
+			if(containsInMonthFormat(ftype))
+			{
+				StringInfoData buf;
+				initStringInfo(&buf);
+				appendStringInfoString(&buf, field[0]);
+				for(int i=1;i<3;i++)
+				{
+					appendStringInfoString(&buf, "-");
+					appendStringInfoString(&buf, field[i]);
+
+				}
+				field[0]=buf.data;
+				ftype[0] = DTK_DATE;
+				nf = nf - 2;
+				for(int i = 1; i < nf; i++)
+				{
+					field[i] = field[i+2];
+					ftype[i] = ftype[i+2];
+				}
+			}
+		}
+
+		for(int i=0;i < nf; i++)
+		{
+			switch(ftype[i])
+			{
+				case DTK_NUMBER:
+					switch(strlen(field[i]))
+					{
+						case 1:
+						case 2:
+							/*
+							 * At this point the 1,2 digit number should be considered as
```

Comment:
are we sure about this? any 1/2 digit number must only come for AM/PM, there are no other cases?

Reply:
Yes, I've tested in SQL Server with different types of formats included.
NOTE: There are cases where input can be of `dd mon yyyy hh am` then an pre-handling of numbers(`dd`, `yyyy`) should be careful, which is already being handled earlier.